internal/witness/witness.go

```diff
@@ -152,16 +145,31 @@ func (wg *WitnessGateway) Witness(ctx context.Context, cp []byte) ([]byte, error
 	return sigBlock.Bytes(), errors.Join(PolicyNotSatisfiedErr, err)
 }
 
-// postResponse contains the parts of the witness response needed for logic flow in order to
-// allow testing.
-type postResponse struct {
-	statusCode int
-	body       []byte
-	headers    map[string][]string
+type consistencyFuture func() ([][]byte, error)
+
+// sharedConsistencyProofFetcher is a thread-safe caching wrapper around a proof builder.
+// This is an optimization for the common case where multiple witnesses are used, and all
+// of the witnesses are of the same size, and thus require the same proof.
+type sharedConsistencyProofFetcher struct {
+	pb      *client.ProofBuilder
+	mu      sync.Mutex
+	results map[uint64]consistencyFuture
 }
 
-// postFn wraps http.Post in order to allow for easier testing.
-type postFn func(ctx context.Context, url, body string) (pr postResponse, err error)
+// ConsistencyProof fetches a consistency proof, reusing any results from parallel requests.
+func (pf *sharedConsistencyProofFetcher) ConsistencyProof(ctx context.Context, smaller, larger uint64) ([][]byte, error) {
+	var f consistencyFuture
+	var ok bool
+	pf.mu.Lock()
+	if f, ok = pf.results[smaller]; !ok {
+		f = sync.OnceValues(func() ([][]byte, error) {
+			return pf.pb.ConsistencyProof(ctx, smaller, larger)
+		})
+		pf.results[smaller] = f
+	}
```

Comment:
This keying of the cache will obviously break if anyone _somehow_ called this method with a different sized `larger` than all the other callers.

I suspect that at the moment this probably can't happen(?), but I wonder if it's worth requiring `larger` be part of the `sharedConsistencyProofFetcher` and validated with each call to `ConsistencyProof`, or widening the key to be a string like e.g. `"<smaller>:<larger>"` to defend against manually-effected baldness for future us if the surrounding code materially changes?